packages/devtools_app/lib/src/debugger/debugger_model.dart

```diff
@@ -252,3 +256,62 @@ class Variable extends TreeNode<Variable> {
     return '${boundVar.name} - $value';
   }
 }
+
+/// A node in a tree of scripts.
+///
+/// A node can either be a directory (a name with potentially some child nodes),
+/// a script reference (where [scriptRef] is non-null), or a combination of both
+/// (where the node has a non-null [scriptRef] but also contains child nodes).
+class FileNode extends TreeNode<FileNode> {
+  FileNode(this.name);
+
+  final String name;
+
+  // This can be null.
+  ScriptRef scriptRef;
+
+  /// This exists to allow for O(1) lookup of children when building the tree.
+  Map<String, FileNode> _childrenAsMap = {};
+
+  bool get hasScript => scriptRef != null;
+
+  /// Given a flat list of service protocol scripts, return a tree of scripts
+  /// representing the best hierarchical grouping.
+  static List<FileNode> createRootsFrom(List<ScriptRef> scripts) {
+    // The name of this node is not exposed to users.
+    final root = FileNode('<root>');
+
+    for (var script in scripts) {
+      final parts = script.uri.split('/');
```

Comment:
what about google3 uris? e.g. `package:foo.bar.baz/src/foobar.dart`

Here's a method in vm_snapshot_analysis that handles this, though it doesn't seem like depending on vm_snapshot_analysis from the debugger makes much sense, so we could just copy it over to devtools:

```
List<String> partsForPath(String path) {
  final parts = path.split('/');
  if (parts.first.startsWith('package:')) {
    // Convert dot separated package name into a path from which this package originated.
    parts.replaceRange(0, 1, parts.first.split('.'));
  }
  return parts;
}
```

vm_snapshot_analysis commit where I added this: https://github.com/dart-lang/sdk/commit/99a233505f5ae4102f6537425e625fe8960c5cef

Reply:
Good point - this should handle google3 urls as well.

Currently it would display them in a flat list at the top level:

```
package:foo.bar.baz/
package:foo.bar.qux/
...
```

I assume it would be better to display them heirarichally, but might need to actually play with it to be sure:

```
package:foo/
  bar/
    baz/
    qux/
...
```

Reply:
cc @grouma who also brought this up

Reply:
+1 for hierarchal.

Reply:
done - we handle dotted paths now